Создание Series

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
my_index = ["Usa", "Canda", "Mexico"]
my_data = [1776, 1867, 1821]
#сделал именованные индексы и данные в объект list

In [ ]:
my_ser = pd.Series(data=my_data, index = my_index)
my_ser
#Создали объект Series

Usa       1776
Canda     1867
Mexico    1821
dtype: int64

In [ ]:
my_ser[0]
my_ser["Usa"]
#можно обращаться к данным по индексу или по именованному индексу

1776

In [ ]:
ages = {"Sam": 5, "Frank": 10, "Spike": 7}
pd.Series(ages)
#создание объекта Series на основе словаря. Индексы уже именованные

Sam       5
Frank    10
Spike     7
dtype: int64

Операции над Series

In [ ]:
q1 = {'Japan': 80, 'China': 450, 'India': 200, 'USA': 250}
q2 = {'Brazil': 100,'China': 500, 'India': 210,'USA': 260}

In [ ]:
sales_q1 = pd.Series(q1)
sales_q2 = pd.Series(q2)

In [ ]:
sales_q1.keys()
#посмотреть ключи объекта

Index(['Japan', 'China', 'India', 'USA'], dtype='object')

In [ ]:
sales_q1 * 2 #все значения умножили на 2 (broadcast)

Japan    160
China    900
India    400
USA      500
dtype: int64

In [ ]:
sales_q1 + sales_q2 #сложили 2 объекта Series, но имеются значения NaN

Brazil      NaN
China     950.0
India     410.0
Japan       NaN
USA       510.0
dtype: float64

In [ ]:
sales_q1.add(sales_q2, fill_value = 0 ) #сложили 2 объекта Series, заменили NaN на 0

Brazil    100.0
China     950.0
India     410.0
Japan      80.0
USA       510.0
dtype: float64

Так же при создании pd.Series(тут можно указать dtype(опрдеелнный тип данных, который нам нужен))

Создание DataFrame

In [ ]:
np.random.seed(101)
my_data = np.random.randint(0,101,(4,3)) #создали двумерный массив 4 на 3

In [ ]:
my_data
my_index = ["CA", "NY", "AZ", "TX"] #создали свои индексы, соответствуют числу строк
my_columns = ['Jan', "Feb", "Mar"] #создали свои названия столбцов, соответствуют числу столбцов

In [ ]:
df = pd.DataFrame(data = my_data, index = my_index, columns = my_columns) #создали DataFrame на основе наших данных, индексов и столбцов
df

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [ ]:
df = pd.read_csv('tips.csv') #считали загруженный файл

In [ ]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

Операции с DataFrame

In [ ]:
df.columns #информация о столбиках(признаках)

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [ ]:
df.index #если индексы именованные - получим их список, если обычные - объект index

RangeIndex(start=0, stop=244, step=1)

In [ ]:
df.head() #по умолчанию выводит первые пять строк, можно указать хоть сколько(показывает с начала)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [ ]:
df.tail() #по умолчанию выводит первые пять строк, можно указать хоть сколько(показывает с конца)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

          Payer Name         CC Number Payment ID  
239    Michael Avila  5296068606052842    Sat2657  
240   Monica Sanders  3506806155565404    Sat1766  
241       Keith Wong  6011891618747196    Sat3880  
242     Dennis Dixon     4375220550950      Sat17  
243  Michelle Hardin  3511451626698139    Thur672

In [ ]:
df.info() #показывает всю информацию о DataFrame

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   total_bill        244 non-null    float64
 1   tip               244 non-null    float64
 2   sex               244 non-null    object 
 3   smoker            244 non-null    object 
 4   day               244 non-null    object 
 5   time              244 non-null    object 
 6   size              244 non-null    int64  
 7   price_per_person  244 non-null    float64
 8   Payer Name        244 non-null    object 
 9   CC Number         244 non-null    int64  
 10  Payment ID        244 non-null    object 
dtypes: float64(3), int64(2), object(6)
memory usage: 21.1+ KB


In [ ]:
df.describe() #применить метрики к числовым столбцам

total_bill         tip        size  price_per_person     CC Number
count  244.000000  244.000000  244.000000        244.000000  2.440000e+02
mean    19.785943    2.998279    2.569672          7.888197  2.563496e+15
std      8.902412    1.383638    0.951100          2.914234  2.369340e+15
min      3.070000    1.000000    1.000000          2.880000  6.040679e+10
25%     13.347500    2.000000    2.000000          5.800000  3.040731e+13
50%     17.795000    2.900000    2.000000          7.255000  3.525318e+15
75%     24.127500    3.562500    3.000000          9.390000  4.553675e+15
max     50.810000   10.000000    6.000000         20.270000  6.596454e+15

count - все значения в таблице, не считая null
>
mean - среднее значение
>
std - стандратное отклонение
>
min - минимальное значение
>
max - максимальное значение
>
25% 50% 75% - квартили, показывающие распределение данных между минимумом и максимумом
>
25% - 25% всех данных меньше указанного числа
>
50% - медиана. 50% счетов были меньше 17.79, а другие 50% были больше
>













In [ ]:
df.describe().transpose() #таблицу можно транспонировать(поменять местами строки и столбцы)

count          mean           std           min  \
total_bill        244.0  1.978594e+01  8.902412e+00  3.070000e+00   
tip               244.0  2.998279e+00  1.383638e+00  1.000000e+00   
size              244.0  2.569672e+00  9.510998e-01  1.000000e+00   
price_per_person  244.0  7.888197e+00  2.914234e+00  2.880000e+00   
CC Number         244.0  2.563496e+15  2.369340e+15  6.040679e+10   

                           25%           50%           75%           max  
total_bill        1.334750e+01  1.779500e+01  2.412750e+01  5.081000e+01  
tip               2.000000e+00  2.900000e+00  3.562500e+00  1.000000e+01  
size              2.000000e+00  2.000000e+00  3.000000e+00  6.000000e+00  
price_per_person  5.800000e+00  7.255000e+00  9.390000e+00  2.027000e+01  
CC Number         3.040731e+13  3.525318e+15  4.553675e+15  6.596454e+15

Операции над столбиками

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [ ]:
df['total_bill'] #вывели только один конкретный столбик

0      16.99
1      10.34
2      21.01
3      23.68
4      24.59
       ...  
239    29.03
240    27.18
241    22.67
242    17.82
243    18.78
Name: total_bill, Length: 244, dtype: float64

In [ ]:
my_cols = ['total_bill', 'tip']
df[my_cols] #вывели конкретные столбики в 2 этапа

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [ ]:
df[['total_bill', 'tip']] #вывели конкретные столбики в 1 этап

total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

In [ ]:
df['tip'] + df['total_bill'] #сложили 2 конкретных столбика

0      18.00
1      12.00
2      24.51
3      26.99
4      28.20
       ...  
239    34.95
240    29.18
241    24.67
242    19.57
243    21.78
Length: 244, dtype: float64

In [ ]:
df['tip_percentage'] = 100 * df['tip'] / df['total_bill'] #создали новый столбец
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [ ]:
df['price_per_person'] = df['total_bill'] / df['size'] #данный столбец уже создан, поэтому просто обновили в нем данные
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [ ]:
df['price_per_person'] = np.round(df['total_bill'] / df['size'], 2) #изменили количество знаков после запятой у столбца
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [ ]:
df.drop('tip_percentage',axis = 1) #удалили столбец. Не вносит измеения в исходный DataFrame

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

axis = 0 - удалить строку, axis = 1 - удалить столбик

In [ ]:
df = df.drop('tip_percentage', axis = 1) #удалили столбец в исходном Data Frame (рекомендованный способ)

Есть еще варинт удалить столбец:
>
df.drop('tip_percentage, axis = 1, inplace = True) (Нерекомендованный способ)


In [ ]:
df.shape
#сначала идут строки, затем столбцы. Поэтому axis = 0 - строки, а axis = 1 - столбцы

(244, 11)

Операции над строчками

In [ ]:
df.set_index('Payment ID') #сделали значения конкретного столбика индексами(не вносит изменения в исходный DataFrame)

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun2959          16.99  1.01  Female     No   Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[244 rows x 10 columns]

In [ ]:
df = df.set_index('Payment ID') #сделали значения конкретного столбика индексами(вносит изменения в исходный DataFrame)

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [ ]:
#df = df.reset_index() #возвращает стандартные индексы

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [ ]:
df.iloc[0] #вывели первую строчку по численному индексу

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
df.loc['Sun2959'] #вывели первую строчку по именованному индексу

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [ ]:
df.iloc[0:4] #вывели строчки с первой по 4, не включая 4

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [ ]:
df.loc[['Sun2959', 'Sun5260']] #вывели 2 строчки по именованному индексу

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun5260                11.84    Nathaniel Harris  4676137647685994

In [ ]:
df = df.drop('Sun2959',axis = 0) #удалили конкретную строчку

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number  
Payment ID                                                        
Sun4608                 3.45    Douglas Tucker  4478071379779230  
Sun4458                 7.00    Travis Walters  6011812112971322  
Sun5260                11.84  Nathaniel Harris  4676137647685994  
Sun2251                 6.15      Tonya Carter  4832732618637221  
Sun9679                 6.32        Erik Smith   213140353657882

In [ ]:
df.iloc[1:] #оставили все строчки кроме первой

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No   Sun  Dinner     4   
Sun5985           8.77  2.00    Male     No   Sun  Dinner     2   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221  
Sun9679                 6.32          Erik Smith   213140353657882  
Sun5985                 4.38  Kristopher Johnson  2223727524230344  
...                      ...                 ...               ...  
Sat2657                 9.68       Michael Avila  5296068606052842  
Sat1766                13.59      Monica Sanders  3506806155565404  
Sat3880                11.34          Keith Wong  6011891618747196  
Sat17                   8.91        Dennis Dixon     4375220550950  
Thur672                 9.39     Michelle Hardin  3511451626698139  

[242 rows x 10 columns]

In [ ]:
one_row = df.iloc[0]

In [ ]:
df = df.append(one_row) #добавили строчку

<ipython-input-48-f5220e0eb126>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(one_row) #добавили строчку


Этот метод скоро будет удален, поэтому нужно будет использовать
>
pd.concat([df,pd.DataFrame([one_row])], axis = 0)
>
Можно вставить строку с уже имеющимся индексом



In [ ]:
df = pd.read_csv('tips.csv')

In [ ]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

Выбор строк по условию 2 способа

In [ ]:
#1 способ. Сумма счета больше 40
bool_series = df['total_bill'] > 40
df[bool_series]

total_bill    tip     sex smoker   day    time  size  price_per_person  \
59        48.27   6.73    Male     No   Sat  Dinner     4             12.07   
95        40.17   4.73    Male    Yes   Fri  Dinner     4             10.04   
102       44.30   2.50  Female    Yes   Sat  Dinner     3             14.77   
142       41.19   5.00    Male     No  Thur   Lunch     5              8.24   
156       48.17   5.00    Male     No   Sun  Dinner     6              8.03   
170       50.81  10.00    Male    Yes   Sat  Dinner     3             16.94   
182       45.35   3.50    Male    Yes   Sun  Dinner     3             15.12   
184       40.55   3.00    Male    Yes   Sun  Dinner     2             20.27   
197       43.11   5.00  Female    Yes  Thur   Lunch     4             10.78   
212       48.33   9.00    Male     No   Sat  Dinner     4             12.08   

          Payer Name         CC Number Payment ID  
59       Brian Ortiz  6596453823950595    Sat8139  
95     Aaron Bentley   180026611638690    Fri9628  
102    Heather Cohen   379771118886604    Sat6240  
142     Eric Andrews  4356531761046453   Thur3621  
156    Ryan Gonzales  3523151482063321    Sun7518  
170    Gregory Clark  5473850968388236    Sat1954  
182     Jose Parsons  4112207559459910    Sun2337  
184      Stephen Cox  3547798222044029    Sun5140  
197      Brooke Soto  5544902205760175   Thur9313  
212  Alex Williamson      676218815212    Sat4590

In [ ]:
#2 способ. Сумма счета больше 40
df[df["total_bill"] > 40]

total_bill    tip     sex smoker   day    time  size  price_per_person  \
59        48.27   6.73    Male     No   Sat  Dinner     4             12.07   
95        40.17   4.73    Male    Yes   Fri  Dinner     4             10.04   
102       44.30   2.50  Female    Yes   Sat  Dinner     3             14.77   
142       41.19   5.00    Male     No  Thur   Lunch     5              8.24   
156       48.17   5.00    Male     No   Sun  Dinner     6              8.03   
170       50.81  10.00    Male    Yes   Sat  Dinner     3             16.94   
182       45.35   3.50    Male    Yes   Sun  Dinner     3             15.12   
184       40.55   3.00    Male    Yes   Sun  Dinner     2             20.27   
197       43.11   5.00  Female    Yes  Thur   Lunch     4             10.78   
212       48.33   9.00    Male     No   Sat  Dinner     4             12.08   

          Payer Name         CC Number Payment ID  
59       Brian Ortiz  6596453823950595    Sat8139  
95     Aaron Bentley   180026611638690    Fri9628  
102    Heather Cohen   379771118886604    Sat6240  
142     Eric Andrews  4356531761046453   Thur3621  
156    Ryan Gonzales  3523151482063321    Sun7518  
170    Gregory Clark  5473850968388236    Sat1954  
182     Jose Parsons  4112207559459910    Sun2337  
184      Stephen Cox  3547798222044029    Sun5140  
197      Brooke Soto  5544902205760175   Thur9313  
212  Alex Williamson      676218815212    Sat4590

In [ ]:
df[df['sex'] == 'Male'] #только мужчины

total_bill   tip   sex smoker  day    time  size  price_per_person  \
1         10.34  1.66  Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50  Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31  Male     No  Sun  Dinner     2             11.84   
5         25.29  4.71  Male     No  Sun  Dinner     4              6.32   
6          8.77  2.00  Male     No  Sun  Dinner     2              4.38   
..          ...   ...   ...    ...  ...     ...   ...               ...   
236       12.60  1.00  Male    Yes  Sat  Dinner     2              6.30   
237       32.83  1.17  Male    Yes  Sat  Dinner     2             16.42   
239       29.03  5.92  Male     No  Sat  Dinner     3              9.68   
241       22.67  2.00  Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75  Male     No  Sat  Dinner     2              8.91   

             Payer Name         CC Number Payment ID  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
5            Erik Smith   213140353657882    Sun9679  
6    Kristopher Johnson  2223727524230344    Sun5985  
..                  ...               ...        ...  
236       Matthew Myers  3543676378973965    Sat5032  
237        Thomas Brown  4284722681265508    Sat2929  
239       Michael Avila  5296068606052842    Sat2657  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  

[157 rows x 11 columns]

In [ ]:
df[df['size'] > 3] #количество гостей больше 3

total_bill   tip     sex smoker   day    time  size  price_per_person  \
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
5         25.29  4.71    Male     No   Sun  Dinner     4              6.32   
7         26.88  3.12    Male     No   Sun  Dinner     4              6.72   
11        35.26  5.00  Female     No   Sun  Dinner     4              8.82   
13        18.43  3.00    Male     No   Sun  Dinner     4              4.61   
23        39.42  7.58    Male     No   Sat  Dinner     4              9.86   
25        17.81  2.34    Male     No   Sat  Dinner     4              4.45   
31        18.35  2.50    Male     No   Sat  Dinner     4              4.59   
33        20.69  2.45  Female     No   Sat  Dinner     4              5.17   
44        30.40  5.60    Male     No   Sun  Dinner     4              7.60   
47        32.40  6.00    Male     No   Sun  Dinner     4              8.10   
52        34.81  5.20  Female     No   Sun  Dinner     4              8.70   
54        25.56  4.34    Male     No   Sun  Dinner     4              6.39   
56        38.01  3.00    Male    Yes   Sat  Dinner     4              9.50   
59        48.27  6.73    Male     No   Sat  Dinner     4             12.07   
63        18.29  3.76    Male    Yes   Sat  Dinner     4              4.57   
77        27.20  4.00    Male     No  Thur   Lunch     4              6.80   
85        34.83  5.17  Female     No  Thur   Lunch     4              8.71   
95        40.17  4.73    Male    Yes   Fri  Dinner     4             10.04   
116       29.93  5.07    Male     No   Sun  Dinner     4              7.48   
119       24.08  2.92  Female     No  Thur   Lunch     4              6.02   
125       29.80  4.20  Female     No  Thur   Lunch     6              4.97   
141       34.30  6.70    Male     No  Thur   Lunch     6              5.72   
142       41.19  5.00    Male     No  Thur   Lunch     5              8.24   
143       27.05  5.00  Female     No  Thur   Lunch     6              4.51   
153       24.55  2.00    Male     No   Sun  Dinner     4              6.14   
154       19.77  2.00    Male     No   Sun  Dinner     4              4.94   
155       29.85  5.14  Female     No   Sun  Dinner     5              5.97   
156       48.17  5.00    Male     No   Sun  Dinner     6              8.03   
157       25.00  3.75  Female     No   Sun  Dinner     4              6.25   
159       16.49  2.00    Male     No   Sun  Dinner     4              4.12   
160       21.50  3.50    Male     No   Sun  Dinner     4              5.38   
167       31.71  4.50    Male     No   Sun  Dinner     4              7.93   
180       34.65  3.68    Male    Yes   Sun  Dinner     4              8.66   
183       23.17  6.50    Male    Yes   Sun  Dinner     4              5.79   
185       20.69  5.00    Male     No   Sun  Dinner     5              4.14   
187       30.46  2.00    Male    Yes   Sun  Dinner     5              6.09   
197       43.11  5.00  Female    Yes  Thur   Lunch     4             10.78   
204       20.53  4.00    Male    Yes  Thur   Lunch     4              5.13   
207       38.73  3.00    Male    Yes   Sat  Dinner     4              9.68   
211       25.89  5.16    Male    Yes   Sat  Dinner     4              6.47   
212       48.33  9.00    Male     No   Sat  Dinner     4             12.08   
216       28.15  3.00    Male    Yes   Sat  Dinner     5              5.63   
219       30.14  3.09  Female    Yes   Sat  Dinner     4              7.54   
227       20.45  3.00    Male     No   Sat  Dinner     4              5.11   
230       24.01  2.00    Male    Yes   Sat  Dinner     4              6.00   

                Payer Name         CC Number Payment ID  
4             Tonya Carter  4832732618637221    Sun2251  
5               Erik Smith   213140353657882    Sun9679  
7              Robert Buck  3514785077705092    Sun8157  
11            Diane Macias  4577817359320969    Sun6686  
13            Joshua Jones  6011163105616890    

Выбор строк по нескольким условиям
>
and заменяется на &
>
or заменяется на |

In [ ]:
df[(df['total_bill'] > 30) & (df['sex'] == 'Male')] #выбрали строки с суммой счета больше 30 и платили мужчины

total_bill    tip   sex smoker   day    time  size  price_per_person  \
23        39.42   7.58  Male     No   Sat  Dinner     4              9.86   
39        31.27   5.00  Male     No   Sat  Dinner     3             10.42   
44        30.40   5.60  Male     No   Sun  Dinner     4              7.60   
47        32.40   6.00  Male     No   Sun  Dinner     4              8.10   
56        38.01   3.00  Male    Yes   Sat  Dinner     4              9.50   
59        48.27   6.73  Male     No   Sat  Dinner     4             12.07   
83        32.68   5.00  Male    Yes  Thur   Lunch     2             16.34   
95        40.17   4.73  Male    Yes   Fri  Dinner     4             10.04   
112       38.07   4.00  Male     No   Sun  Dinner     3             12.69   
141       34.30   6.70  Male     No  Thur   Lunch     6              5.72   
142       41.19   5.00  Male     No  Thur   Lunch     5              8.24   
156       48.17   5.00  Male     No   Sun  Dinner     6              8.03   
167       31.71   4.50  Male     No   Sun  Dinner     4              7.93   
170       50.81  10.00  Male    Yes   Sat  Dinner     3             16.94   
173       31.85   3.18  Male    Yes   Sun  Dinner     2             15.92   
175       32.90   3.11  Male    Yes   Sun  Dinner     2             16.45   
179       34.63   3.55  Male    Yes   Sun  Dinner     2             17.32   
180       34.65   3.68  Male    Yes   Sun  Dinner     4              8.66   
182       45.35   3.50  Male    Yes   Sun  Dinner     3             15.12   
184       40.55   3.00  Male    Yes   Sun  Dinner     2             20.27   
187       30.46   2.00  Male    Yes   Sun  Dinner     5              6.09   
207       38.73   3.00  Male    Yes   Sat  Dinner     4              9.68   
210       30.06   2.00  Male    Yes   Sat  Dinner     3             10.02   
212       48.33   9.00  Male     No   Sat  Dinner     4             12.08   
237       32.83   1.17  Male    Yes   Sat  Dinner     2             16.42   

                Payer Name         CC Number Payment ID  
23          Lance Peterson  3542584061609808     Sat239  
39       Mr. Brandon Berry  6011525851069856    Sat6373  
44             Todd Cooper      503846761263    Sun2274  
47            James Barnes  3552002592874186    Sun9677  
56   James Christensen DDS   349793629453226    Sat8903  
59             Brian Ortiz  6596453823950595    Sat8139  
83           Daniel Murphy  5356177501009133   Thur8801  
95           Aaron Bentley   180026611638690    Fri9628  
112             Jeff Lopez  3572865915176463     Sun591  
141         Steven Carlson  3526515703718508   Thur1025  
142           Eric Andrews  4356531761046453   Thur3621  
156          Ryan Gonzales  3523151482063321    Sun7518  
167         Michael Lawson  3566285921227119    Sun3719  
170          Gregory Clark  5473850968388236    Sat1954  
173            Scott Perez  3577115550328507    Sun9335  
175        Nathan Reynolds   370307040837149    Sun5109  
179           Brian Bailey   346656312114848    Sun9851  
180       James Hebert DDS      676168737648    Sun7544  
182           Jose Parsons  4112207559459910    Sun2337  
184            Stephen Cox  3547798222044029    Sun5140  
187          David Barrett  4792882899700988    Sun9987  
207          Ricky Ramirez   347817964484033    Sat4505  
210          Shawn Mendoza    30184049218122    Sat8361  
212        Alex Williamson      676218815212    Sat4590  
237           Thomas Brown  4284722681265508    Sat2929

In [ ]:
df[(df['day'] == 'Sat') | (df['day'] == 'Sun')] #выбрали чеки в субботу или воскресенье

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0         16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No  Sun  Dinner     4              6.15   
..          ...   ...     ...    ...  ...     ...   ...               ...   
238       35.83  4.67  Female     No  Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No  Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes  Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No  Sat  Dinner     2              8.91   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
238      Kimberly Crane      676184013727    Sat9777  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  

[163 rows x 11 columns]

Хотим добавить еще несколько дней
>
Есть два способа
>
1) Добавить условие в строку кода выше
>
2) Метод isin() - True, если есть в списке значений


In [ ]:
df[df['day'].isin(['Sat','Sun'])] #можем добавлять значения в список

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0         16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No  Sun  Dinner     4              6.15   
..          ...   ...     ...    ...  ...     ...   ...               ...   
238       35.83  4.67  Female     No  Sat  Dinner     3             11.94   
239       29.03  5.92    Male     No  Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes  Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes  Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No  Sat  Dinner     2              8.91   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
238      Kimberly Crane      676184013727    Sat9777  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  

[163 rows x 11 columns]